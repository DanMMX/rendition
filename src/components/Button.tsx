import * as React from 'react';
import styled, { StyledFunction, withTheme } from 'styled-components';
import hoc from '../hoc';
import { normal, bold, darken, getColor, getColoringType, px } from '../utils';

interface ButtonProps extends DefaultProps, Coloring, Sizing {
	square?: boolean;
	disabled?: boolean;
	outline?: boolean;
	plaintext?: boolean;
	underline?: boolean;
}

const squareWidth = (val: number): number => val / 9 * 10;

const minWidth = (props: ButtonProps) => {
	if (props.w == null && !props.square) {
		return 'auto';
	}
	if (props.square) {
		return props.emphasized
			? squareWidth(props.theme.space[5])
			: squareWidth(props.theme.space[4]);
	}

	return props.w;
};

const horizontalPadding = (props: ButtonProps) => {
	if (props.w == null && !props.square) {
		return props.emphasized ? 50 : 20;
	}
	if (props.square || props.w) {
		return 0;
	}

	return props.emphasized ? 50 : 20;
};

const styleableButton: StyledFunction<ButtonProps> = styled.button;

const Button = styleableButton`
	padding-top: 1px;
	padding-left: ${props => px(horizontalPadding(props))};
	padding-right: ${props => px(horizontalPadding(props))};
	font-family: inherit;
	display: inline-block;
	font-weight: ${props => bold(props)};
	border-radius: ${props => px(props.theme.radius)};
	appearance: none;
	text-decoration: none;
	border: 0;
	margin: 0;
	min-width: ${props => px(minWidth(props))};
	vertical-align: middle;
	font-size: inherit;
	line-height: 1.1;
	text-align: center;
	cursor: pointer;
	background: ${props => props.bg || getColor(props, 'bg', 'main')};
	color: ${props => props.color || '#fff'};
	height: ${(props: ButtonProps) =>
		px(props.emphasized ? props.theme.space[5] : props.theme.space[4])};

	&:hover,
	&:focus,
	&:active {
		color: #fff;
		background-color: ${props =>
			getColor(props, 'bg', 'dark') || darken(props.color as string)};
	}

	&:disabled {
		opacity: 0.65;
		cursor: not-allowed;
	}
`;

const Outline = Button.extend`
	color: ${props =>
		getColor(props, 'color', 'main') || props.theme.colors.text.main};
	background: ${props => props.color || 'none'};
	border: 1px solid;
`;

const Plaintext = Button.extend`
	padding-left: 0;
	padding-right: 0;
	height: auto;
	border: 0;
	border-radius: 0;
	color: ${props =>
		getColor(props, 'color', 'main') || props.theme.colors.text.main};
	background: ${props => props.color || 'none'};
	font-weight: ${props => normal(props)};

	&:hover,
	&:focus,
	&:active {
		background: none;
		color: ${props =>
			getColor(props, 'color', 'dark') || props.theme.colors.text.main};
	}
`

const Underline = Plaintext.extend`
	padding-bottom: 2px;
	border-bottom: 1px solid;
	font-weight: ${props => bold(props)};

	&:hover,
	&:focus,
	&:active {
		color: ${props =>
			getColor(props, 'color', 'main') || props.theme.colors.text.main};
		box-shadow: 0px -1px 0 0px inset;
	}
`;

export default withTheme(
	hoc(({ outline, underline, plaintext, ...props }: ButtonProps) => {
		if (plaintext) {
			return <Plaintext {...props} />;
		} else if (outline) {
			return <Outline {...props} />;
		} else if (underline) {
			return <Underline {...props} />;
		} else if (!getColoringType(props) && !props.color && !props.bg) {
			// outline tertiary is our default btn
			return <Outline {...props} tertiary />;
		} else {
			return <Button {...props} />;
		}
	}),
);
