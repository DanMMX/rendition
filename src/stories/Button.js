import * as React from 'react'
import { storiesOf, action } from '@storybook/react'
import styled from 'styled-components'
import * as FaExpand from 'react-icons/lib/fa/expand'
import * as FaRecycle from 'react-icons/lib/fa/recycle'
import * as FaSpinner from 'react-icons/lib/fa/spinner'
import { Button } from '../'

const Container = styled.div`
  margin: 30px;
`

storiesOf('Button', module)
  .addWithInfo('Standard', () => {
    return (
      <Container>
        <Button mx={2} primary onPress={action('case-1')}>
          Button
        </Button>
        <Button mx={2} secondary onPress={action('case-2')}>
          Button
        </Button>
        <Button mx={2} success onPress={action('case-3')}>
          Button
        </Button>
        <Button mx={2} tertiary onPress={action('case-4')}>
          Button
        </Button>
        <Button mx={2} onPress={action('case-5')}>
          Button
        </Button>
      </Container>
    )
  })
  .addWithInfo('Emphasized', () => {
    return (
      <Container>
        <Button mx={2} emphasized primary onPress={action('case-1')}>
          Button
        </Button>
        <Button mx={2} emphasized secondary onPress={action('case-2')}>
          Button
        </Button>
        <Button mx={2} emphasized success onPress={action('case-3')}>
          Button
        </Button>
        <Button mx={2} emphasized tertiary onPress={action('case-4')}>
          Button
        </Button>
        <Button mx={2} emphasized onPress={action('case-5')}>
          Button
        </Button>
      </Container>
    )
  })
  .addWithInfo('Disabled', () => {
    return (
      <Container>
        <Button mx={2} disabled primary onPress={action('case-1')}>
          Button
        </Button>
        <Button mx={2} disabled secondary onPress={action('case-2')}>
          Button
        </Button>
        <Button mx={2} disabled success onPress={action('case-3')}>
          Button
        </Button>
        <Button mx={2} disabled tertiary onPress={action('case-4')}>
          Button
        </Button>
        <Button mx={2} disabled onPress={action('case-5')}>
          Button
        </Button>
      </Container>
    )
  })
  .addWithInfo('Outline', () => {
    return (
      <Container>
        <Button mx={2} primary outline onPress={action('case-1')}>
          Button
        </Button>
        <Button mx={2} danger outline onPress={action('case-3')}>
          Button
        </Button>
        <Button mx={2} tertiary outline onPress={action('case-3')}>
          Button
        </Button>
        <Button mx={2} outline onPress={action('case-4')}>
          Button
        </Button>
      </Container>
    )
  })
  .addWithInfo('Underline', () => {
    return (
      <Container>
        <Button mx={2} primary underline onPress={action('case-1')}>
          Button
        </Button>
        <Button mx={2} underline onPress={action('case-4')}>
          Button
        </Button>
      </Container>
    )
  })
  .addWithInfo('Plaintext', () => {
    return (
      <Container>
        <Button mx={2} plaintext primary onPress={action('case-1')}>
          Button
        </Button>
        <Button
          mx={2}
          plaintext
          primary
          iconElement={<FaSpinner />}
          onPress={action('case-2')}
        >
          Button
        </Button>
      </Container>
    )
  })
  .addWithInfo('Width', () => {
    return (
      <Container>
        <Button w={95} mx={2} primary onPress={action('case-1')}>
          Button
        </Button>
        <Button w={60} mx={2} secondary onPress={action('case-2')}>
          Button
        </Button>
        <Button w={120} mx={2} tertiary onPress={action('case-3')}>
          Button
        </Button>
        <div style={{ width: 300 }}>
          <Button w='100%' m={2} onPress={action('case-4')}>
            Button
          </Button>
        </div>
      </Container>
    )
  })
  .addWithInfo('Square', () => {
    return (
      <Container>
        <Button mx={2} tertiary square onPress={action('case-3')}>
          <FaExpand />
        </Button>
        <Button m={2} square onPress={action('case-4')}>
          <FaRecycle />
        </Button>
      </Container>
    )
  })
  .addWithInfo('Square Emphasized', () => {
    return (
      <Container>
        <Button mx={2} emphasized tertiary square onPress={action('case-3')}>
          <FaExpand />
        </Button>
        <Button m={2} emphasized square onPress={action('case-4')}>
          <FaRecycle />
        </Button>
      </Container>
    )
  })
