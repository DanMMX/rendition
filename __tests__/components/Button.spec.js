import React from 'react';
import renderer from 'react-test-renderer';
import Provider from '../../src/components/Provider'
import Button from '../../src/components/Button'

test('Button renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Button>Click me</Button>
    </Provider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
