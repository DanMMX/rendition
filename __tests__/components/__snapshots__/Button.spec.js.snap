// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <button
    className="sc-bZQynM bycwzs sc-bxivhb ierOrI"
  >
    Click me
  </button>
</div>
`;
